lsp-mode.el

```diff
@@ -1859,42 +1852,63 @@ WORKSPACE is the workspace that contains the progress token."
-         (-map (-lambda ((code count))
-                 (propertize (format "%s" count)
-                             'face (cl-case code
-                                     ('error 'error)
-                                     ('warning 'warning)
-                                     ('info 'success)
-                                     ('hint 'success)))))
-         (s-join "/")
-         (format "%s"))))
+                       ((equal :file lsp-diagnostics-modeline-scope)
+                        (list (lsp--get-buffer-diagnostics)))
+                       (t (->> (eq :workspace lsp-diagnostics-modeline-scope)
+                               (lsp-diagnostics)
+                               (ht-values))))))
+    (->> (let ((stats (make-vector lsp/diagnostic-severity-max 0))
+               strs
+               (i 0))
+           (mapc (lambda (buf-diags)
+                   (mapc (lambda (diag)
+                           (-let [(&Diagnostic :severity?) diag]
+                             (cl-incf (aref stats (or severity? 0)))))
+                         buf-diags))
+                 diagnostics)
+           (while (< i lsp/diagnostic-severity-max)
+             (when (> (aref stats i) 0)
+               (setq strs
+                     (nconc strs
+                            `(,(propertize
+                                (format "%s" (aref stats i))
+                                'face
+                                (cond
+                                  ((equal i lsp/diagnostic-severity-error) 'error)
+                                  ((equal i lsp/diagnostic-severity-warning) 'warning)
+                                  ((equal i lsp/diagnostic-severity-information) 'success)
+                                  ((equal i lsp/diagnostic-severity-hint) 'success)))))))
+             (cl-incf i))
+           strs)
+         (s-join "/"))))
+
+(defun lsp--diagnostics-update-modeline ()
+  "Update diagnostics modeline string."
+  (let ((str (lsp--diagnostics-modeline-statistics)))
+    (setq lsp--diagnostics-modeline-string
+          (if (string-empty-p str) ""
+            (concat " " str)))))
 
 (define-minor-mode lsp-diagnostics-modeline-mode
   "Toggle diagnostics modeline."
   :group 'lsp-mode
   :global nil
   :lighter ""
-  (let ((status '(t (:eval (concat " " (lsp--diagnostics-modeline-statistics) " ")))))
-    (setq-local global-mode-string
-                (cond ((and lsp-diagnostics-modeline-mode
-                            (not (-contains? global-mode-string status)))
-                       (cons status global-mode-string))
-                      (t (remove status global-mode-string))))))
+  (cond
+    (lsp-diagnostics-modeline-mode
+     (add-to-list 'global-mode-string '(t (:eval lsp--diagnostics-modeline-string)))
+     (add-hook 'lsp-after-open-hook 'lsp--diagnostics-update-modeline nil t)
+     (add-hook 'lsp-diagnostics-updated-hook 'lsp--diagnostics-update-modeline nil t))
```

Comment:
this would mean that for all of the open buffers in lsp-mode we will update the lsp--diagnostics-modeline-string, right? 

Can we change the code like that:

1. When there is diagnostics update lsp--diagnostics-modeline-string is set to nil.
2. When modeline is built it either uses lsp--diagnostics-modeline-string if present or builts it and caches it.

WDYT?

Reply:
This one is local hook so it will run once when buffer is opened and every time diagnostic is updated. `lsp--diagnostics-modeline-string` is also a buffer local variable too.

When `lsp-diagnostics-modeline-scope` is `:file`, it's pretty optimal and we need to calculate it anyway.
When `lsp-diagnostics-modeline-scope` is `:workspace`, we're wasting recalculating the modeline for every opened buffers in the workspace.
When `lsp-diagnostics-modeline-scope` is `:global`, we're wasting recalculating the modeline for all opened lsp buffers.

We can try using cache to reduce that. But I would expected the number of opened buffers will not be that much.